### ISSUE TITLE
Restructure Roles

### ISSUE THREAD

**Issue by kdmnk** [EotvosCollegium/mars#392]
> I was also thinking on making the roles time constrained. That way, we could have regular collegists, plus an additional resident role that is set to be valid for 1 year/indefinite. We can also use that for guests, instead of the guest_until attribute and solve #291. (Also, the general collegist role is also valid only until the end of their studies by default which we can set this way).
> 
> @machiato32 what do you think, does putting effort into this would worth it? (We have some time to work on this, the earliest this will be relevant is next summer)

> I'm starting to like this idea. The role_user table would have a created_at and valid_until field, and we would use a global scope to filter on those. This way, we don't delete roles just invalidate them, making a log out of the table.

_Originally posted by @kdmnk in https://github.com/EotvosCollegium/mars/issues/358#issuecomment-1873828995_

**Comment by viktorcsimma** [comment]
I think I'll take this up; it's probably going to be important soon.

**Comment by kdmnk** [comment]
Nice. One more thing: does it make sense to have a boolean flag for stating that a role is indefinite or time constrained? If it is time constrained the validity has to be set when the role is attached to the user

**Comment by viktorcsimma** [comment]
I don't think it's necessary; it could simply be indicated by the `valid_until` field being `null`.

**Comment by viktorcsimma** [comment]
Hm... but the new resident role could be either indefinite or time-constrained, so I don't know how that could be implemented. I would rather omit that for now.

**Comment by viktorcsimma** [comment]
No, you are right; we need it for the UI. It would be odd for each and every role to put a date picker next to the name:D

> Hm... but the new resident role could be either indefinite or time-constrained, so I don't know how that could be implemented. I would rather omit that for now.
